### PR TITLE
refactor: add useMemo & useCallback

### LINKED ISSUE #41680: 4.x/5.x表格渲染性能优化：internalRowClassName、expandIcon 需要用 useMemo 缓存
### What problem does this feature solve?

表格数据100条时操作卡顿问题

### What does the proposed API look like?

修改 `Table.tsx`:
https://github.com/ant-design/ant-design/blob/307b76b478fe701b91197ecc19ce835e26fb397e/components/table/Table.tsx#L438

```typescript
  const internalRowClassName = React.useMemo((record: RecordType, index: number, indent: number) => {
    let mergedRowClassName: string;
    if (typeof rowClassName === 'function') {
      mergedRowClassName = classNames(rowClassName(record, index, indent));
    } else {
      mergedRowClassName = classNames(rowClassName);
    }

    return classNames(
      {
        [`${prefixCls}-row-selected`]: selectedKeySet.has(getRowKey(record, index)),
      },
      mergedRowClassName,
    );
  }, [ rowClassName, selectedKeySet, getRowKey ]);

  // ========================== Expandable ==========================

  // Pass origin render status into `rc-table`, this can be removed when refactor with `rc-table`
  (mergedExpandable as any).__PARENT_RENDER_ICON__ = mergedExpandable.expandIcon;

  // Customize expandable icon
  mergedExpandable.expandIcon = React.useMemo(() => {
    return mergedExpandable.expandIcon || expandIcon || renderExpandIcon(tableLocale!);
  }, [ mergedExpandable.expandIcon, expandIcon, tableLocale ]);
```

### PR DESCRIPTION
**Pull request by li-jia-nan**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- close #41680

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | refactor: add useMemo & useCallback |
| 🇨🇳 Chinese | - |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 0cb3698</samp>

Improve table component performance and readability with hooks and types. Use `React.useCallback` and `React.useMemo` to avoid unnecessary re-rendering and re-computation, and import `RenderExpandIcon` type from `rc-table` to simplify the code in `components/table/InternalTable.tsx`.

### <samp>🤖 Generated by Copilot at 0cb3698</samp>

*  Import `RenderExpandIcon` type from `rc-table` library to define custom expandable icon for table rows ([link](https://github.com/ant-design/ant-design/pull/41683/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R5))
*  Memoize `internalRowClassName` function with `React.useCallback` hook to improve performance and avoid unnecessary re-rendering of table rows ([link](https://github.com/ant-design/ant-design/pull/41683/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616L418-R431))
*  Memoize `mergedExpandable.expandIcon` assignment with `React.useMemo` hook to improve performance and avoid unnecessary re-rendering of expandable icons ([link](https://github.com/ant-design/ant-design/pull/41683/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616L440-R441))

### CODE REVIEW

**Comment by afc163** [comment]
加个用例避免改挂？

File: components/table/InternalTable.tsx [hunk 1]
```diff
@@ -415,30 +413,29 @@ const InternalTable = <RecordType extends AnyObject = any>(
...
-    }
-
-    return classNames(
-      {
-        [`${prefixCls}-row-selected`]: selectedKeySet.has(getRowKey(record, index)),
-      },
-      mergedRowClassName,
-    );
-  };
+  const internalRowClassName = React.useCallback(
```
Comment (zombieJ) [inline components/table/InternalTable.tsx:416]: 这个 useCallback 没有意义。因为还是要执行代码，这边 memo 一下反而使得渲染多了一次对比 deps 的损耗。

File: components/table/InternalTable.tsx [hunk 2]
```diff
@@ -415,30 +413,29 @@ const InternalTable = <RecordType extends AnyObject = any>(
...
 
   // ========================== Expandable ==========================
 
   // Pass origin render status into `rc-table`, this can be removed when refactor with `rc-table`
   (mergedExpandable as any).__PARENT_RENDER_ICON__ = mergedExpandable.expandIcon;
 
   // Customize expandable icon
-  mergedExpandable.expandIcon =
-    mergedExpandable.expandIcon || expandIcon || renderExpandIcon(tableLocale!);
+  mergedExpandable.expandIcon = React.useMemo<RenderExpandIcon<RecordType>>(
```
Comment (zombieJ) [inline components/table/InternalTable.tsx:435]: 同上，除非走到 `renderExpandIcon`，否则原来的性能会更好

**Comment by li-jia-nan** [comment]
先 close 啦

**Pull request closed without merging**